source/slang/slang-ir-constexpr.cpp

```diff
@@ -343,154 +344,175 @@ bool propagateConstExprForward(PropagateConstExprContext* context, IRGlobalValue
-                                UInt paramIndex = paramCounter++;
+    IRBuilder builder(context->getModule());
+    builder.setInsertInto(code);
 
-                                auto param = pp;
-                                auto arg = callInst->getOperand(firstCallArg + paramIndex);
+    bool anyChanges = false;
+    for (;;)
+    {
+        // Note: we are walking the list of blocks and the instructions
+        // in each block in reverse order, to maximize the chances that
+        // we propagate multiple changes in a each pass.
+        //
+        // TODO: this should probably all be done with a work list instead,
+        // but that requires being able to detect instructions vs. other
+        // values.
 
-                                if (isConstExpr(param))
-                                {
-                                    if (maybeMarkConstExprBackwardPass(context, arg))
-                                    {
-                                        changedThisIteration = true;
-                                    }
-                                }
-                            }
-                        }
-                    }
-                    else
-                    {
-                        // If we don't have a concrete callee function
-                        // definition, then we need to extract the
-                        // type of the callee instruction, and try to work
-                        // with that.
-                        //
-                        // Note that this does not allow us to propagate
-                        // `constexpr` information from the body of a callee
-                        // back to call sites.
-                        auto calleeType = callee->getDataType();
-                        if (auto caleeFuncType = as<IRFuncType>(calleeType))
-                        {
-                            auto paramCount = caleeFuncType->getParamCount();
-                            for (UInt pp = 0; pp < paramCount; ++pp)
-                            {
-                                auto paramType = caleeFuncType->getParamType(pp);
-                                auto arg = callInst->getOperand(firstCallArg + pp);
-                                if (isConstExpr(paramType))
-                                {
-                                    if (maybeMarkConstExprBackwardPass(context, arg))
-                                    {
-                                        changedThisIteration = true;
-                                    }
-                                }
-                            }
-                        }
-                    }
-                }
+        bool changedThisIteration = false;
+        for (auto bb = code->getLastBlock(); bb; bb = bb->getPrevBlock())
+        {
+            for (auto ii = bb->getLastInst(); ii; ii = ii->getPrevInst())
+            {
+                changedThisIteration = checkInstConstExprRecursively(context, ii);
```

Comment:
Same here, `changedThisIteration` may be reset back to false if a later iteration is false.